### ISSUE TITLE
Deployable Output analysis

### ISSUE THREAD

**Issue by snorfalorpagus** [snorfalorpagus/pywr#7]
Implement calculation of conjunctive-use deployable output.

* English method
* Scottish method

**Comment by jetuk** [comment]
Should be tackled by the wider issue of multiple runs discussed in #24 ?

**Comment by snorfalorpagus** [comment]
@jetuk Yes and no. Running a model with different parameters in parallel will be useful for calculating DO, but it's not the whole story.

For the Scottish method parallel runs will speed things up greatly, as every run provides some useful information. Once all of the results have come in, they still need to be turned into an overall curve. Aquator does this by fitting to a Generalised Extreme Value distribution. I think MISER only does the English method. It looks like there are some GEV functions in scipy which we can use: http://docs.scipy.org/doc/scipy-0.13.0/reference/generated/scipy.stats.genextreme.html

For the English method running models in parallel is complicated. As soon as a model fails we know that any DO above that is (or at least should be!) a failure. I'm not sure how much is gained by running models in parallel, vs using a bisection search.

You're right that the demand scaling could probably be achieved using the "scenario" code in #24.

**Comment by jetuk** [comment]
@snorfalorpagus OK, but I think they should both be implemented as an application which uses a `ModelController` to achieve multiple runs. Whether those runs are sequential, in parallel or both shouldn't matter if the controller can do it and report relevant metrics.

While bisection is simple in a single thread there might be some marginal improvement by doing with a few in parallel.

**Comment by snorfalorpagus** [comment]
I agree we need a controller to manage the runs.

Here's the original paper describing the "scottish method": http://nora.nerc.ac.uk/9093/1/Low_Flow_01.pdf

We need to track the number of failures in a model run.

**Comment by snorfalorpagus** [comment]
For the scottish method #170 covers one half of the problem (event tracking). The other half is just running for different demands, which is something already possible using scenarios.

**Comment by jetuk** [comment]
I was thinking about this again.

Is this a good summary.

### Scottish method
- Multiple runs at different demands.
- Each run is to completion. 
- Track the number of failure events in each run.
- Use above to estimate GEV distribution

Therefore a script could be written as follows,

- Input
  - base model JSON file.
  - range of demand (min, max, number of steps)
  - list of output nodes to use in the analysis
  - list of nodes to track for failure (storage or output? + failure threshold)
- Run
  - Add a `Scenario` with an associated `Parameter` to vary demand at the specified nodes.
  - Add event tracking recorder at the specified nodes.
  - Run all scenarios together.
  - Save event statistics.
  - Use `scipy.stats` to fit a GEV (this is possible a post-processing step)

We could define JSON spec to the define the Scottish method. So you would then do,

```
python scottish_method.py myzone_setup.json myzone_model.json
```
### English method
- Search for the maximum demand that causes no failures.

This is different to the above because you need to search. I guess bisection is good. And it would be relatively easy to search multiple scenarios together (e.g. for climate change analysis).


So do we have a separate script for the two methods or try to combine?

**Comment by snorfalorpagus** [comment]
Speaking to Doug a little while ago he said the GEV part is a bit questionable anyway (issues with selection of hyperparameters). The event recording is the important bit.

I don't mind having separate scripts for each.

In practical terms - how I we actually modify the demands? I know we can make use of the scenario functionality. But how does the model know which to change? And how *exactly*?

**Comment by jetuk** [comment]
Something like this,

https://gist.github.com/jetuk/817dcffd6e6643581895f892a12f7667

**Comment by snorfalorpagus** [comment]
I was thinking that this was going to be harder because the `node.max_flow` includes the demand savings. But actually these are just a factor anyway and it doesn't matter what order you do the multiplication in. Nice work.

**Comment by jetuk** [comment]
So should we bundle something like this in Pywr? I guess the actual analysis could be abstracted to a library function with the script doing very little. We would need a new sub-package (e.g. `pywr.utils`).

The above doesn't cope with scenarios. That would need some work.
